### DISCUSSION TITLE
Unpoly server side protocol and redirects

### DISCUSSION THREAD

**Discussion by apollo13** [unpoly/unpoly#260] [General]
Hi,

I am not opening this as question or as issue, because it is not an issue per se and there might not actually be an answer. I am implementing the server-side protocol in Python and looked at the rails implementation for inspiration (or bluntly put I copied most of the ideas there ;)). 

When looking at the redirect handling I noticed that the rails variant copies all request headers over to GET parameters. Looking at the dev tools in my browser (latest Firefox) it seems that the request headers `X-Up-*` are also sent for the redirect (even when the method changes from POST to GET for instance). This makes me wonder if one could optimize this by only sending the parameters which the server can change / actually changed via GET parameters.

I tried to find any documentation on which headers browsers are required to pass in the case of a redirect but I couldn't find any. Do you have any idea on whether it is guaranteed that the browser will send the initial headers to a redirect with a `XMLHttpRequest` that unpoly seems to use?

Thanks,
Florian

**Comment by triskweline** [comment]
Persisting `X-Up` headers over redirects was not included in the protocol due to the complexity of a server-side implementation. `unpoly-rails` could be half the code without it.

Some people have since expressed interest in this becoming part of the protocol:

- https://github.com/unpoly/unpoly/issues/213
- https://github.com/unpoly/unpoly/issues/214

> This makes me wonder if one could optimize this by only sending the parameters which the server can change / actually changed via GET parameters.

There are benefits to persisting all headers. E.g. when `/foo` redirects to `/bar`, the code handling `/bar` may want to know which fragment was originally targeted (`X-Up-Target`), or whether we're rendering content in an overlay (`X-Up-Mode`).

> I tried to find any documentation on which headers browsers are required to pass in the case of a redirect but I couldn't find any. Do you have any idea on whether it is guaranteed that the browser will send the initial headers to a redirect with a XMLHttpRequest that unpoly seems to use?

I'm afraid I didn't get what you mean? When you make a request through Unpoly, it will send the `X-Up` request headers documented in the [protocol](https://unpoly.com/up.protocol). When the server responds with a redirect, the browser will automatically make a new request to the new URL. For that second request no `X-Up` headers will be sent, and it is impossible for client-side JavaScript to modify that second request.

**Reply by apollo13** [reply]
> There are benefits to persisting all headers. ... may want to know which fragment was originally targeted

I agree, but see below

> When the server responds with a redirect, the browser will automatically make a new request to the new URL. For that second request no X-Up headers will be sent, and it is impossible for client-side JavaScript to modify that second request.

This is not what I am seeing though; I am making a request to `/modal_redirect/` which issues a 302 to `/modal/` and the headers are present:
```
{'Content-Length': '', 'Content-Type': 'text/plain', 'Host': 'localhost:8000', 'User-Agent': 'Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0', 'Accept': '*/*', 'Accept-Language': 'en-US,en;q=0.7,de-DE;q=0.3', 'Accept-Encoding': 'gzip, deflate', 'Referer': 'http://localhost:8000/', 'X-Up-Target': '[up-main~=modal]', 'X-Up-Fail-Target': 'main', 'X-Up-Mode': 'modal', 'X-Up-Fail-Mode': 'root', 'X-Up-Context': '{"test":"asd"}', 'X-Up-Fail-Context': '{}', 'X-Up-Version': '2.2.1', 'Dnt': '1', 'Connection': 'keep-alive', 'Sec-Fetch-Dest': 'empty', 'Sec-Fetch-Mode': 'cors', 'Sec-Fetch-Site': 'same-origin', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache'}
[24/Aug/2021 06:26:22] "GET /modal_redirect/ HTTP/1.1" 302 0
{'Content-Length': '', 'Content-Type': 'text/plain', 'Host': 'localhost:8000', 'User-Agent': 'Mozilla/5.0 (X11; Fedora; Linux x86_64; rv:91.0) Gecko/20100101 Firefox/91.0', 'Accept': '*/*', 'Accept-Language': 'en-US,en;q=0.7,de-DE;q=0.3', 'Accept-Encoding': 'gzip, deflate', 'X-Up-Target': '[up-main~=modal]', 'X-Up-Fail-Target': 'main', 'X-Up-Mode': 'modal', 'X-Up-Fail-Mode': 'root', 'X-Up-Context': '{"test":"asd"}', 'X-Up-Fail-Context': '{}', 'X-Up-Version': '2.2.1', 'Referer': 'http://localhost:8000/', 'Dnt': '1', 'Connection': 'keep-alive', 'Sec-Fetch-Dest': 'empty', 'Sec-Fetch-Mode': 'cors', 'Sec-Fetch-Site': 'same-origin', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache'}
[24/Aug/2021 06:26:22] "GET /modal/?_up_version=2.2.1&_up_mode=modal&_up_context=%7B%22test%22%3A+%22asd%22%7D&_up_fail_target=main&_up_fail_mode=root&_up_fail_context=%7B%7D&_up_title=Testing+33 HTTP/1.1" 200 910
```

The prints above the log lines are a simple print of `request.headers`. Matching devtool screenshot from request 1:

![Screenshot from 2021-08-24 08-30-02](https://user-images.githubusercontent.com/81547/130567478-9e4b7f9f-3a6d-46e8-90cd-94e3d8b094f1.png)
and request 2:
![Screenshot from 2021-08-24 08-30-14](https://user-images.githubusercontent.com/81547/130567490-035bace2-2e43-49ed-8780-f4e016249e2f.png)

I just tested on Chrome, same behavior… Honestly it kinda makes sense given that a redirect is apparently absolutely transparent for javascript. I know you just said this wouldn't happen, but do you mind trying this on your end? I would have a massive bug here if it worked for me but not for you.

**Comment by triskweline** [comment]
I can confirm the browsers re-send request headers over redirects (tested in Chrome and IE11).

I too spent some time looking for a spec that documents this behavior, but to no avail :disappointed: Maybe this is an implementation quirk that browsers kept copying for backwards compatibility.

I found [this integration test](https://github.com/WebKit/WebKit/blob/8afe31a018b11741abdf9b4d5bb973d7c1d9ff05/LayoutTests/http/tests/xmlhttprequest/redirections-and-user-headers.html) for WebKit.

> This makes me wonder if one could optimize this by only sending the parameters which the server can change / actually changed via GET parameters.

I now agree that this is a good optimization to make.

**Reply by apollo13** [reply]
Same for Curl, I wonder if we can find something there that hints to a spec or something.
